controllers/request_controller.go

```diff
@@ -0,0 +1,462 @@
+		statusPatch.SetWaitingForIssuerReadyOutdated()
+
+		return result, statusPatch, nil // apply patch, done
+	}
+	if readyCondition.Status != cmmeta.ConditionTrue {
+		logger.V(1).Info("Issuer is not Ready yet (status == false). Waiting for it to become ready.", "issuer ready condition", readyCondition)
+		statusPatch.SetWaitingForIssuerReadyNotReady(readyCondition)
+
+		return result, statusPatch, nil // apply patch, done
+	}
+
+	signedCertificate, err := r.Sign(log.IntoContext(ctx, logger), requestObjectHelper.RequestObject(), issuerObject)
+	if err == nil {
+		logger.V(1).Info("Successfully finished the reconciliation.")
+		statusPatch.SetIssued(signedCertificate)
+
+		return result, statusPatch, nil // apply patch, done
+	}
+
+	// An error in the issuer part of the operator should trigger a reconcile
+	// of the issuer's state.
+	if issuerError := new(signer.IssuerError); errors.As(err, issuerError) {
+		if reportError := r.EventSource.ReportError(
+			issuerGvk, client.ObjectKeyFromObject(issuerObject),
+			issuerError.Err,
+		); reportError != nil {
+			return result, nil, fmt.Errorf("unexpected ReportError error: %v", reportError) // requeue with backoff
+		}
+
+		logger.V(1).Info("Issuer is not Ready yet (ready condition out-of-date). Waiting for it to become ready.", "issuer-error", issuerError)
+		statusPatch.SetWaitingForIssuerReadyOutdated()
+
+		return result, statusPatch, nil // apply patch, done
+	}
+
+	didCustomConditionTransition := false
+	if targetCustom := new(signer.SetCertificateRequestConditionError); errors.As(err, targetCustom) {
+		logger.V(1).Info("Set RequestCondition error. Setting condition.", "error", err)
+		didCustomConditionTransition = statusPatch.SetCustomCondition(
+			string(targetCustom.ConditionType),
+			metav1.ConditionStatus(targetCustom.Status),
+			targetCustom.Reason,
+			targetCustom.Error(),
+		)
+	}
+
+	// Check if we have still time to requeue & retry
+	isPendingError := errors.As(err, &signer.PendingError{})
+	isPermanentError := errors.As(err, &signer.PermanentError{})
+	pastMaxRetryDuration := r.Clock.Now().After(requestObject.GetCreationTimestamp().Add(r.MaxRetryDuration))
+	if !isPendingError && (isPermanentError || pastMaxRetryDuration) {
+		// fail permanently
+		logger.V(1).Error(err, "Permanent Request error. Marking as failed.")
+		statusPatch.SetPermanentError(err)
+
+		return result, statusPatch, reconcile.TerminalError(err) // apply patch, done
+	} else {
+		// retry
+		logger.V(1).Error(err, "Retryable Request error.")
+		statusPatch.SetRetryableError(err)
+
+		if didCustomConditionTransition {
+			// the reconciliation loop will be retriggered because of the added/ changed custom condition
+			return result, statusPatch, reconcile.TerminalError(err) // apply patch, done
+		} else {
+			return result, statusPatch, err // apply patch, requeue with backoff
+		}
+	}
+}
```

Comment:
**Note:** Let's refactor this as part of #59. There should be three code paths here, not just two ("in progress and all is well", "error that will be retried", and "terminal error").